### PR TITLE
Code refactoring for ranking objective & Faster ndcg_xendcg

### PR DESCRIPTION
**Pull request by guolinke**
1. optimize the O(n^2) loop in `ndcg_xendcg` to O(n).
2. Fix non-thread-safe random generator in `ndcg_xendcg`, which should not share a `Random` object in multi-threading. Furthermore, the random results should be identical now regards to different num_threads.

Also some code refactorings.

Benchmark:

Boosting cost:
| Data      | Master|  This PR |
|----|  ----| ----| 
| Yahoo LTR | 9.36 s |1.45 s |
| MS LTR|72.02 s | 2.72 s|

### CODE REVIEW

File: src/objective/rank_objective.hpp [hunk 1]
```diff
@@ -1,68 +1,131 @@
 /*!
  * Copyright (c) 2016 Microsoft Corporation. All rights reserved.
```
Comment (guolinke) [inline src/objective/rank_objective.hpp]:
```suggestion
 * Copyright (c) 2020 Microsoft Corporation. All rights reserved.
```

**Comment by guolinke** [comment]
ping @sbruch for the changes in `ndcg_xendcg` part

File: tests/python_package_test/test_sklearn.py [hunk 2]
```diff
@@ -129,7 +129,7 @@ def test_xendcg(self):
                 eval_metric='ndcg',
                 callbacks=[lgb.reset_parameter(learning_rate=lambda x: max(0.01, 0.1 - 0.01 * x))])
         self.assertLessEqual(gbm.best_iteration_, 24)
-        self.assertGreater(gbm.best_score_['valid_0']['ndcg@1'], 0.6559)
+        self.assertGreater(gbm.best_score_['valid_0']['ndcg@1'], 0.6382)
         self.assertGreater(gbm.best_score_['valid_0']['ndcg@3'], 0.6421)
```
Comment (guolinke) [inline tests/python_package_test/test_sklearn.py]:
```suggestion
        self.assertGreater(gbm.best_score_['valid_0']['ndcg@3'], 0.6319)
```

**Comment by guolinke** [comment]
@sbruch I will merge this first. If there are any problems, feel free to open the issue/PR.

**Comment by StrikerRUS** [comment]
@guolinke This PR removes old params related to ranking objective. Do you think it is not "breaking"?

**Comment by guolinke** [comment]
@StrikerRUS oh, I forget that change.

**Pull request merged**